File: neetcode-150/arrays-hashing/encode-and-decode-strings/solution.java
import java.util.*;

/**
 * Time Complexity: O(n) - Where n is total characters
 * Space Complexity: O(1) - Excluding input/output
 */
public class Codec {
    // Encodes a list of strings to a single string.
    public String encode(List<String> strs) {
        StringBuilder encoded = new StringBuilder();
        
        for (String str : strs) {
            encoded.append(str.length()).append('#').append(str);
        }
        
        return encoded.toString();
    }

    // Decodes a single string to a list of strings.
    public List<String> decode(String s) {
        List<String> decoded = new ArrayList<>();
        int i = 0;
        
        while (i < s.length()) {
            // Find the delimiter '#'
            int delimiterIndex = s.indexOf('#', i);
            
            // Extract length
            int length = Integer.parseInt(s.substring(i, delimiterIndex));
            
            // Extract string
            String str = s.substring(delimiterIndex + 1, delimiterIndex + 1 + length);
            decoded.add(str);
            
            // Move to next string
            i = delimiterIndex + 1 + length;
        }
        
        return decoded;
    }
}

// Alternative approach using more explicit parsing
class CodecAlternative {
    public String encode(List<String> strs) {
        StringBuilder sb = new StringBuilder();
        
        for (String str : strs) {
            sb.append(str.length()).append("#").append(str);
        }
        
        return sb.toString();
    }
    
    public List<String> decode(String s) {
        List<String> result = new ArrayList<>();
        int i = 0;
        
        while (i < s.length()) {
            int j = i;
            while (s.charAt(j) != '#') {
                j++;
            }
            
            int length = Integer.parseInt(s.substring(i, j));
            String str = s.substring(j + 1, j + 1 + length);
            result.add(str);
            
            i = j + 1 + length;
        }
        
        return result;
    }
}

// Using streams for more functional approach
class CodecStreams {
    public String encode(List<String> strs) {
        return strs.stream()
                .map(str -> str.length() + "#" + str)
                .reduce("", String::concat);
    }
    
    public List<String> decode(String s) {
        List<String> result = new ArrayList<>();
        int i = 0;
        
        while (i < s.length()) {
            int delimiterIndex = s.indexOf('#', i);
            int length = Integer.parseInt(s.substring(i, delimiterIndex));
            String str = s.substring(delimiterIndex + 1, delimiterIndex + 1 + length);
            result.add(str);
            i = delimiterIndex + 1 + length;
        }
        
        return result;
    }
}

File: neetcode-150/arrays-hashing/encode-and-decode-strings/solution.js
/**
 * Time Complexity: O(n) - Where n is total characters
 * Space Complexity: O(1) - Excluding input/output
 */
var encode = function(strs) {
    let encoded = '';
    
    for (const str of strs) {
        encoded += str.length + '#' + str;
    }
    
    return encoded;
};

var decode = function(s) {
    const decoded = [];
    let i = 0;
    
    while (i < s.length) {
        // Find the delimiter '#'
        const delimiterIndex = s.indexOf('#', i);
        
        // Extract length
        const length = parseInt(s.substring(i, delimiterIndex));
        
        // Extract string
        const str = s.substring(delimiterIndex + 1, delimiterIndex + 1 + length);
        decoded.push(str);
        
        // Move to next string
        i = delimiterIndex + 1 + length;
    }
    
    return decoded;
};

// Alternative approach using more explicit parsing
var encodeAlternative = function(strs) {
    return strs.map(str => str.length + '#' + str).join('');
};

var decodeAlternative = function(s) {
    const result = [];
    let i = 0;
    
    while (i < s.length) {
        let j = i;
        while (s[j] !== '#') {
            j++;
        }
        
        const length = parseInt(s.substring(i, j));
        const str = s.substring(j + 1, j + 1 + length);
        result.push(str);
        
        i = j + 1 + length;
    }
    
    return result;
};

// Using reduce for more functional approach
var encodeFunctional = function(strs) {
    return strs.reduce((acc, str) => acc + str.length + '#' + str, '');
};

var decodeFunctional = function(s) {
    const result = [];
    let i = 0;
    
    while (i < s.length) {
        const delimiterIndex = s.indexOf('#', i);
        const length = parseInt(s.substring(i, delimiterIndex));
        const str = s.substring(delimiterIndex + 1, delimiterIndex + 1 + length);
        result.push(str);
        i = delimiterIndex + 1 + length;
    }
    
    return result;
};
